### PR TITLE
Shopping: Provider Dropdown 

### LINKED ISSUE #11265: Shopping: Provider Dropdown
## Feature Description

For shopping we want to implement a dropdown to choose between existing providers, like we do for Monetization:

Screenshot 2022-04-14 at 11 55 55

**Note:** Exact design and wording is still TBD, but we can already do a lot of the work here. Especially since it's behind a flag anyway.

The options for the dropdown would be:

* None
* WooCommerce
* Shopify

### Implementation Brief

* Register a new setting in `Settings` class (e.g. `web_stories_shopping_provider`)
	* Can be an enum with `none`, `woocommerce`, `shopify`, similar to the `web_stories_ad_network` setting
* Add new dropdown on the settings page for choosing the shopping provider
	* Can reuse a lot of code & tests from the monetization settings section
* Add e2e tests (can reuse code from the monetization e2e test)

### PR DESCRIPTION
**Pull request by timarney**
## Context

For shopping we want to implement a dropdown to choose between existing providers

see: https://github.com/GoogleForCreators/web-stories-wp/issues/11265#issue-1204331478

## Summary

Adds shopping provider dropdown with the following options:
- None
- WooCommerce
- Shopify

provider

## Relevant Technical Choices

Note: I'm also updated the code to show or hide Shopify fields based on the provider selected.

## Testing Instructions

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1. Ensure the shopping feature is enabled
2. Visit settings
3. Input the shopping provider 
4. Reload the page to ensure the dropdown reloads with the correct value

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [ ] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [ ] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #11265

### CODE REVIEW

File: includes/Settings.php [hunk 1]
```diff
@@ -275,6 +280,17 @@ public function register(): void {
 			]
 		);
 
+		register_setting(
+			self::SETTING_GROUP,
+			self::SETTING_NAME_SHOPPING_PROVIDER,
+			[
+				'description'  => __( 'Shopping provider', 'web-stories' ),
+				'type'         => 'string',
+				'default'      => '',
```
Comment (spacedmonkey) [inline includes/Settings.php]:
```suggestion
				'default'      => 'none',
				'enum'         => [ 'none', 'woocommerce', 'shopify' ],
```

Let's follow the pattern from ad network.

File: packages/e2e-tests/src/specs/dashboard/settings/adminUser/shoppingProvider.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+ */
+import {
+  visitSettings,
+  withExperimentalFeatures,
+} from '@web-stories-wp/e2e-test-utils';
+
+/**
+ * Internal dependencies
+ */
+import { shoppingProviderDropdownSelector } from '../../../../utils';
```
Comment (spacedmonkey) [inline packages/e2e-tests/src/specs/dashboard/settings/adminUser/shoppingProvider.js]: Does this need to in util directly. It is used elsewhere?

File: packages/e2e-tests/src/specs/dashboard/settings/adminUser/shoppingProvider.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+describe('Admin User', () => {
```
Comment (spacedmonkey) [inline packages/e2e-tests/src/specs/dashboard/settings/adminUser/shoppingProvider.js]: Let's have a better description for this. I know this done other places, but these should changed as well.

**Comment by spacedmonkey** [comment]
Should we also change these lines. 

https://github.com/GoogleForCreators/web-stories-wp/blob/e40da3898e793d41ba7ece07f8bde578737712ae/includes/REST_API/Products_Controller.php#L146-L156

**Comment by spacedmonkey** [comment]
Screenshot 2022-04-25 at 09 33 36

When nothing selected, this should default to `none`.

File: packages/wp-dashboard/src/components/editorSettings/shopping/shoppingProviderDropDown.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+ */
+import { SHOPPING_PROVIDER_TYPE } from '../../../constants';
+
+export const TEXT = {
+  SLOT_ID_LABEL: __('Shopping provider', 'web-stories'),
+};
+
+const OPTIONS = [
+  {
+    label: _x('None', 'ad network', 'web-stories'),
```
Comment (swissspidy) [inline packages/wp-dashboard/src/components/editorSettings/shopping/shoppingProviderDropDown.js]:
```suggestion
    label: _x('None', 'shopping provider', 'web-stories'),
```

File: packages/wp-dashboard/src/components/editorSettings/shopping/shoppingProviderDropDown.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  },
+  {
+    label: __('WooCommerce', 'web-stories'),
+    value: SHOPPING_PROVIDER_TYPE.WOOCOMMERCE,
+  },
+  {
+    label: __('Shopify', 'web-stories'),
+    value: SHOPPING_PROVIDER_TYPE.SHOPIFY,
+  },
+];
```
Comment (spacedmonkey) [inline packages/wp-dashboard/src/components/editorSettings/shopping/shoppingProviderDropDown.js:46]: @swissspidy Do you want to make this filterable? As in, allow developers to add their own options here?

**Comment by timarney** [comment]
> Should we also change these lines.

@spacedmonkey --- wasn't sure if you might pick that up in the Product query ticket.

I'll add it to this PR.

File: packages/e2e-tests/src/specs/dashboard/settings/adminUser/shoppingProvider.js (hunk 2 above)
Comment (timarney) [reply packages/e2e-tests/src/specs/dashboard/settings/adminUser/shoppingProvider.js]: Following the `monetizationDropdownSelector` ---  but that dropdown also has tests for author etc... I'll pull it into the test file direct for now.

**Comment by spacedmonkey** [comment]
> wasn't sure if you might pick that up in the Product query ticket.

It might be, I can always rebase from this PR.

File: packages/wp-dashboard/src/components/editorSettings/shopping/test/shoppingProvider.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+import { screen } from '@testing-library/react';
+
+/**
+ * Internal dependencies
+ */
+import ShoppingProviderDropDown from '..';
+import { SHOPPING_PROVIDER_TYPE } from '../../../../constants';
+import { renderWithProviders } from '../../../../testUtils';
+
+describe('Editor Settings: Shopping provider settings <ShoppingProviderDropDown />', function () {
```
Comment (spacedmonkey) [inline packages/wp-dashboard/src/components/editorSettings/shopping/test/shoppingProvider.js]: Why are we using function and not arrow functions here?

Comment (timarney) [reply packages/wp-dashboard/src/components/editorSettings/shopping/test/shoppingProvider.js]: I'll update --- just some copy paste from 

```
describe('Editor Settings: Ad network settings <AdNetworkSettings />', function () {
```

**Comment by timarney** [comment]
> Screenshot 2022-04-25 at 09 33 36
> 
> When nothing selected, this should default to `none`.

Re-tested resetting my DB --- believe this is fixed now.

File: includes/REST_API/Products_Controller.php [hunk 7]
```diff
@@ -144,15 +144,16 @@ public function get_items( $request ) {
...
-			return $this->get_items_shopify( $request );
+		if( $shopping_provider === 'woocommerce' ){
+			return $this->get_items_woocommerce( $request );
 		}
 
-		if ( function_exists( 'wc_get_products' ) ) {
-			return $this->get_items_woocommerce( $request );
+		if( $shopping_provider === 'shopify' ){
+			$shopify_host         = $this->settings->get_setting( Settings::SETTING_NAME_SHOPIFY_HOST );
+			$shopify_access_token = $this->settings->get_setting( Settings::SETTING_NAME_SHOPIFY_ACCESS_TOKEN );
```
Comment (spacedmonkey) [inline includes/REST_API/Products_Controller.php]: These lines are not needed?

File: includes/REST_API/Products_Controller.php [hunk 8]
```diff
@@ -144,15 +144,16 @@ public function get_items( $request ) {
 		$results = [];
 
 		// TODO(#11154): Refactor to extract product query logic out of this controller.
-		$shopify_host         = $this->settings->get_setting( Settings::SETTING_NAME_SHOPIFY_HOST );
-		$shopify_access_token = $this->settings->get_setting( Settings::SETTING_NAME_SHOPIFY_ACCESS_TOKEN );
+		$shopping_provider = $this->settings->get_setting( Settings::SETTING_NAME_SHOPPING_PROVIDER);
 
-		if ( ! empty( $shopify_host ) && ! empty( $shopify_access_token ) ) {
-			return $this->get_items_shopify( $request );
+		if( $shopping_provider === 'woocommerce' ){
```
Comment (spacedmonkey) [inline includes/REST_API/Products_Controller.php]: Instead of if statements, a switch statement might be cleaner here.

Comment (spacedmonkey) [reply includes/REST_API/Products_Controller.php]: Example 

```
switch ( $shopping_provider ){
    case 'woocommerce':
        $response = $this->get_items_woocommerce( $request );
        break;
     case 'shopify':
       $response = $this->get_items_shopify( $request );
       break;
   default:
     $response = rest_ensure_response( [] );
}

return $response;

Comment (timarney) [reply includes/REST_API/Products_Controller.php]: Sure sound good - updated

File: includes/REST_API/Products_Controller.php (hunk 7 above)
Comment (timarney) [reply includes/REST_API/Products_Controller.php]: Think this is outdated now

File: includes/REST_API/Products_Controller.php [hunk 9]
```diff
@@ -46,193 +46,200 @@
...
+     *
+     * @param WP_REST_Request $request Full details about the request.
+     * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+     */
+    public function get_items($request)
+    {
+        // TODO(#11154): Refactor to extract product query logic out of this controller.
+        $shopping_provider = $this->settings->get_setting(Settings::SETTING_NAME_SHOPPING_PROVIDER);
+
+        switch ( $shopping_provider ){
```
Comment (spacedmonkey) [inline includes/REST_API/Products_Controller.php]: This is soo clean. Love it!

File: packages/wp-dashboard/src/components/editorSettings/shopping/shoppingProviderDropDown.js (hunk 5 above)
Comment (swissspidy) [reply packages/wp-dashboard/src/components/editorSettings/shopping/shoppingProviderDropDown.js:46]: Not at this time

File: includes/REST_API/Products_Controller.php [hunk 10]
```diff
@@ -46,193 +46,200 @@
...
+     * Get the list of service IDs required for this service to be registered.
+     *
+     * Needed because the story post type needs to be registered first.
+     *
+     * @since 1.20.0
+     *
+     * @return string[] List of required services.
+     */
+    public static function get_requirements(): array
+    {
```
Comment (swissspidy) [inline includes/REST_API/Products_Controller.php]: Odd that our PHPCS config does not catch this, but we don't usually put the opening `{` bracket on the next line in our code base. Somehow in this file _all brackets_ are now suddenly on the next line.

If you could fix that, that would be great.

Comment (timarney) [reply includes/REST_API/Products_Controller.php]: That file is fixed up now.

**Review by swissspidy — APPROVED** [review]

**Pull request merged**